feat: Add comprehensive features and security implementation

Unicyclop opened the pull request:
Features:
- Extract MediaUploadDialog into separate component for cleaner code organization
- Implement delete mode with click-to-select individual posts
- Implement edit mode with EditContentDialog component
- Add ability to attach content (text/media/links) when editing posts
- Make bookmarks not auto-pinned (isBookmarked: false by default)
- Make bookmark URLs clickable with proper styling
- Rename 'Save Bookmark' to 'Save Link' throughout UI
- Add 4-button bubble menu (Text Post, Media, Link, Edit Mode)

Security (CWE-434 Protection):
- Create backend/src/utils/secureFileUpload.js with comprehensive validation
- Implement strict whitelist: PNG, JPG, JPEG, GIF, WebP, PDF, TXT, MD
- Add MIME type validation and extension checking
- Implement MIME-extension matching to prevent spoofing
- Add filename sanitization (remove path separators, null bytes, special chars)
- Generate unique filenames: {timestamp}_{random}_{sanitized_name}
- Implement path traversal prevention with double-checking
- Store uploads outside web root (/var/app/uploads)
- Set file permissions to 0o644 (read-only for others)
- Add .htaccess for Apache protection
- Implement UploadLogger class for security monitoring
- Add frontend file validation in MediaUploadDialog and EditContentDialog
- Add server-side validation simulation in api.ts uploadMedia
- Document security implementation in backend/SECURITY.md (400+ lines)

Testing:
- Update FloatingMenu.test.tsx with 20 comprehensive tests
- Test coverage for bubble menu, delete mode, edit mode, content creation

Fixes:
- Remove missing CaptchaType67 import from login component
- Fix FloatingMenu test queries to use queryByRole for dialogs
- Update tests to handle renamed 'Save Link' buttons

Unicyclop committed:
feat: Add comprehensive features and security implementation

Features:
- Extract MediaUploadDialog into separate component for cleaner code organization
- Implement delete mode with click-to-select individual posts
- Implement edit mode with EditContentDialog component
- Add ability to attach content (text/media/links) when editing posts
- Make bookmarks not auto-pinned (isBookmarked: false by default)
- Make bookmark URLs clickable with proper styling
- Rename 'Save Bookmark' to 'Save Link' throughout UI
- Add 4-button bubble menu (Text Post, Media, Link, Edit Mode)

Security (CWE-434 Protection):
- Create backend/src/utils/secureFileUpload.js with comprehensive validation
- Implement strict whitelist: PNG, JPG, JPEG, GIF, WebP, PDF, TXT, MD
- Add MIME type validation and extension checking
- Implement MIME-extension matching to prevent spoofing
- Add filename sanitization (remove path separators, null bytes, special chars)
- Generate unique filenames: {timestamp}_{random}_{sanitized_name}
- Implement path traversal prevention with double-checking
- Store uploads outside web root (/var/app/uploads)
- Set file permissions to 0o644 (read-only for others)
- Add .htaccess for Apache protection
- Implement UploadLogger class for security monitoring
- Add frontend file validation in MediaUploadDialog and EditContentDialog
- Add server-side validation simulation in api.ts uploadMedia
- Document security implementation in backend/SECURITY.md (400+ lines)

Testing:
- Update FloatingMenu.test.tsx with 20 comprehensive tests
- All 120 frontend tests passing
- Test coverage for bubble menu, delete mode, edit mode, content creation

Fixes:
- Remove missing CaptchaType67 import from login component
- Fix FloatingMenu test queries to use queryByRole for dialogs
- Update tests to handle renamed 'Save Link' buttons

DrewProkarym commented on frontend/src/utils/api.ts:
```diff
@@ -192,26 +192,288 @@ export const api = {
+   * Get media for a space (from localStorage)
    */
   async getMedia(spaceId: string) {
-    return apiFetch(`/media?spaceId=${spaceId}`);
```
These should be able to use the updated media routes

DrewProkarym commented on frontend/src/components/MediaUploadDialog.tsx:
```diff
@@ -0,0 +1,309 @@
+                type="file"
+                onChange={handleFileChange}
+                className="hidden"
+                accept=".png,.jpg,.jpeg,.gif,.webp,.pdf,.txt,.md"
```
We currently use
  ".png": ["image/png"],
  ".jpg": ["image/jpeg"],
  ".jpeg": ["image/jpeg"],
  ".gif": ["image/gif"],
  ".webp": ["image/webp"],
  ".bmp": ["image/bmp"],
  ".svg": ["image/svg+xml"],
  ".txt": ["text/plain"],
  ".json": ["application/json"],

Unicyclop committed:
Merge origin/main into Y-Nguyen-floating-menu-update

- Integrated backend media API routes with frontend MediaUploadDialog
- Updated file size limit from 10MB to 20MB
- Updated allowed file types: .png, .jpg, .jpeg, .gif, .webp, .bmp, .svg, .txt, .json
- Replaced localStorage media logic with backend API calls (api.createMedia)
- Resolved merge conflicts in 6 files:
  - App.tsx: Combined search functionality with delete/edit mode features
  - ContentArea.tsx: Preserved pinned/unpinned sections with delete/edit mode
  - ContentCard.tsx: Kept comprehensive delete/edit mode UI
  - FloatingMenu.tsx: Kept delete/edit mode state management
  - login.tsx: Added captcha verify mode from main
  - api.ts: Kept backend media endpoints, retained localStorage for posts/bookmarks
- MediaUploadDialog now converts files to base64 for backend transmission
- All frontend media operations now use backend /media endpoints

Unicyclop committed:
Add backend media integration with Docker volumes

- Added express.static middleware to serve uploaded files
- Increased file size limit to 50MB for media uploads
- Added support for video formats (MP4, WebM, MOV, AVI)
- Fixed database schema: changed file_size from smallint to integer
- Added persistent Docker volume for backend uploads
- Updated frontend to fetch media from backend API
- Fixed space ID mapping for container associations
- All 114 frontend tests passing

DrewProkarym commented on frontend/src/components/FloatingMenu.tsx:
```diff
@@ -196,28 +135,117 @@ export function FloatingMenu({ currentSpaceId, currentUserId, onContentAdded, on
+      
+      localStorage.setItem('ss_posts', JSON.stringify(filteredPosts));
+      localStorage.setItem('ss_media', JSON.stringify(filteredMedia));
+      localStorage.setItem('ss_bookmarks', JSON.stringify(filteredBookmarks));
```
Only reason I'm suggesting this is because variable naming is actually a really helpful skill to practice. It helps a lot more with readability than I used to think.

DrewProkarym commented on frontend/src/components/EditContentDialog.tsx:
```diff
@@ -0,0 +1,437 @@
+  // States for adding attachments
+  const [showAttachments, setShowAttachments] = useState(false);
+  const [attachmentType, setAttachmentType] = useState<'text' | 'image' | 'link' | null>(null);
+  const [attachmentText, setAttachmentText] = useState("");
+  const [attachmentUrl, setAttachmentUrl] = useState("");
+  const [attachmentFile, setAttachmentFile] = useState<File | null>(null);
+  const [attachmentPreview, setAttachmentPreview] = useState<string>("");
+
+  // CWE-434: Security - File upload validation
+  const ALLOWED_EXTENSIONS = ['png', 'jpg', 'jpeg', 'gif', 'webp', 'pdf', 'txt', 'md'];
+  const ALLOWED_MIME_TYPES = [
+    'image/png',
+    'image/jpeg',
+    'image/gif',
+    'image/webp',
+    'application/pdf',
+    'text/plain',
+    'text/markdown'
+  ];
+  const MAX_FILE_SIZE = 10 * 1024 * 1024; // 10MB
+
+  const validateFile = (file: File): { valid: boolean; errors: string[] } => {
+    const errors: string[] = [];
+
+    if (file.size > MAX_FILE_SIZE) {
+      errors.push(`File size must be less than ${MAX_FILE_SIZE / (1024 * 1024)}MB`);
+    }
+
+    const fileName = file.name.toLowerCase();
+    const extension = fileName.split('.').pop() || '';
+    if (!ALLOWED_EXTENSIONS.includes(extension)) {
+      errors.push(`File type not allowed. Allowed: ${ALLOWED_EXTENSIONS.join(', ')}`);
+    }
+
+    if (!ALLOWED_MIME_TYPES.includes(file.type)) {
+      errors.push('File MIME type not allowed');
+    }
+
+    return { valid: errors.length === 0, errors };
+  };
```
fix this please

DrewProkarym commented on frontend/src/components/EditContentDialog.tsx:
```diff
@@ -0,0 +1,437 @@
+  
+  // States for adding attachments
+  const [showAttachments, setShowAttachments] = useState(false);
+  const [attachmentType, setAttachmentType] = useState<'text' | 'image' | 'link' | null>(null);
+  const [attachmentText, setAttachmentText] = useState("");
+  const [attachmentUrl, setAttachmentUrl] = useState("");
+  const [attachmentFile, setAttachmentFile] = useState<File | null>(null);
+  const [attachmentPreview, setAttachmentPreview] = useState<string>("");
+
+  // CWE-434: Security - File upload validation
+  const ALLOWED_EXTENSIONS = ['png', 'jpg', 'jpeg', 'gif', 'webp', 'pdf', 'txt', 'md'];
+  const ALLOWED_MIME_TYPES = [
+    'image/png',
+    'image/jpeg',
+    'image/gif',
+    'image/webp',
+    'application/pdf',
+    'text/plain',
+    'text/markdown'
+  ];
+  const MAX_FILE_SIZE = 10 * 1024 * 1024; // 10MB
+
+  const validateFile = (file: File): { valid: boolean; errors: string[] } => {
+    const errors: string[] = [];
+
+    if (file.size > MAX_FILE_SIZE) {
+      errors.push(`File size must be less than ${MAX_FILE_SIZE / (1024 * 1024)}MB`);
+    }
+
+    const fileName = file.name.toLowerCase();
+    const extension = fileName.split('.').pop() || '';
+    if (!ALLOWED_EXTENSIONS.includes(extension)) {
+      errors.push(`File type not allowed. Allowed: ${ALLOWED_EXTENSIONS.join(', ')}`);
+    }
+
+    if (!ALLOWED_MIME_TYPES.includes(file.type)) {
+      errors.push('File MIME type not allowed');
+    }
+
+    return { valid: errors.length === 0, errors };
```
This is either outdated, wrong, or not necessary for the frontend. Ignore this

DrewProkarym commented on frontend/src/components/__tests__/FloatingMenu.test.tsx:
```diff
@@ -7,77 +7,257 @@ import { vi } from 'vitest';
-        const dialogs = document.querySelectorAll('[role="dialog"]');
-        expect(dialogs.length).toBeGreaterThan(0);
-      }, { timeout: 2000 });
+        const dialogTitle = screen.queryByRole('heading', { name: 'Upload Media' });
```
outdated

DrewProkarym commented on frontend/src/components/__tests__/FloatingMenu.test.tsx:
```diff
@@ -92,78 +272,205 @@ describe('FloatingMenu Component', () => {
+          const createBtn = screen.getByText('Create Post');
+          fireEvent.click(createBtn);
+          
+          waitFor(() => {
+            expect(mockOnContentAdded).toHaveBeenCalled();
           });
+        }
```
Outdated

DrewProkarym commented on frontend/src/components/MediaUploadDialog.tsx:
```diff
@@ -0,0 +1,314 @@
+  const [mediaTitle, setMediaTitle] = useState("");
+  const [mediaDescription, setMediaDescription] = useState("");
+  const [mediaFile, setMediaFile] = useState<File | null>(null);
+  const [mediaPreview, setMediaPreview] = useState<string | null>(null);
+
+  // CWE-434: Security - Whitelist of allowed file types (matches backend)
+  const ALLOWED_EXTENSIONS = ['png', 'jpg', 'jpeg', 'gif', 'webp', 'bmp', 'svg', 'txt', 'json'];
+  const ALLOWED_MIME_TYPES = [
+    'image/png',
+    'image/jpeg', 
+    'image/gif',
+    'image/webp',
+    'image/bmp',
+    'image/svg+xml',
+    'text/plain',
+    'application/json'
+  ];
```
@Unicyclop the file types and maps do not match across secureFileUpload.js, EditContentDialog.tsx, MediaUploadDialog.tsx, and mediaServices.js

DrewProkarym commented on frontend/src/components/MediaUploadDialog.tsx:
```diff
@@ -0,0 +1,314 @@
+                type="file"
+                onChange={handleFileChange}
+                className="hidden"
+                accept=".png,.jpg,.jpeg,.gif,.webp,.bmp,.svg,.txt,.json"
```
the file types and maps do not match across secureFileUpload.js, EditContentDialog.tsx, MediaUploadDialog.tsx, and mediaServices.js

DrewProkarym commented on frontend/src/components/__tests__/FloatingMenu.test.tsx:
```diff
@@ -92,78 +272,205 @@ describe('FloatingMenu Component', () => {
+    render(<FloatingMenu currentSpaceId="test-space" currentUserId="test-user" />);
+    
+    const buttons = screen.getAllByRole('button');
+    const deleteBtn = buttons.find(btn => btn.textContent === '🗑️');
```
@Unicyclop accept this commit

DrewProkarym commented on frontend/src/components/EditContentDialog.tsx:
```diff
@@ -0,0 +1,437 @@
+import { Label } from "./ui/label";
+import { Button } from "./ui/button";
+import { Textarea } from "./ui/textarea";
+import { api } from "../utils/api";
```
@Unicyclop can you verify that this file should not be calling the API to do updates and, if not, remove the unused API import?

DrewProkarym commented on frontend/src/App.tsx:
```diff
@@ -1,12 +1,14 @@
 import { FloatingMenu } from './components/FloatingMenu';
+import { EditContentDialog } from './components/EditContentDialog';
 import AnimatedBackground from './components/AnimatedBackground';
+import { api } from './utils/api';
```
outdated

DrewProkarym commented:
The allowed media upload size is inconsistent across files. Use ctrl+f and search "1024" to see them.

DrewProkarym commented on frontend/src/components/MediaUploadDialog.tsx:
```diff
@@ -0,0 +1,334 @@
+          type: 'image', // Adjust based on file type
+          content: {
+            id: result.data?.id,
+            title: mediaTitle,
+            description: mediaDescription,
```
These columns do not exist on the media table, the backend media model, or in the insertion query. Please add them @Unicyclop

DrewProkarym commented on frontend/src/components/__tests__/FloatingMenu.test.tsx:
```diff
@@ -92,78 +272,205 @@ describe('FloatingMenu Component', () => {
+          setTimeout(() => {
+            const updatedTextPostBtn = screen.queryByTitle('Create Text Post');
+            expect(updatedTextPostBtn).not.toBeInTheDocument();
+          }, 500);
```
Commit this suggestion @Unicyclop

DrewProkarym commented on backend/SECURITY.md:
I am not sure what var/app/uploads you are referring to here but the backend is correct for media stored in the API already. They would be at backend/src/uploads/*/. Were you referring to a frontend storage?

DrewProkarym commented on backend/src/utils/secureFileUpload.js:
```diff
@@ -0,0 +1,451 @@
+
+# Allow only specific safe file types
+<FilesMatch "\\.(?i:png|jpg|jpeg|gif|webp|pdf|txt|md)$">
+  Allow from all
```
Commit this suggestion

DrewProkarym commented on frontend/src/components/FloatingMenu.tsx:
```diff
@@ -196,28 +135,117 @@ export function FloatingMenu({ currentSpaceId, currentUserId, onContentAdded, on
+      setSelectedItems([]);
+      
+      // Reload the page to reflect changes
+      window.location.reload();
```
idk, the second sounds right but use copilot in vs code itself

DrewProkarym commented on frontend/src/App.tsx:
```diff
@@ -31,13 +33,307 @@ export default function App() {
+            contentBySpace[spaceName].push({
+              type: 'image',
+              content: {
+                id: item.id,
+                title: item.title,
+                description: item.description,
+                image: item.fileData,
+                timestamp: new Date(item.createdAt).toLocaleString(),
+                isBookmarked: item.isBookmarked || false
+              }
+            });
+          } else if (item.type === 'bookmark') {
+            let domain = '';
+            try {
+              const urlObj = new URL(item.url);
+              domain = urlObj.hostname;
+            } catch (e) {
+              domain = item.url;
+            }
+            
+            contentBySpace[spaceName].push({
+              type: 'link',
+              content: {
+                id: item.id,
+                title: item.title,
+                text: item.notes || 'No description provided',
+                domain: domain,
+                timestamp: new Date(item.createdAt).toLocaleString(),
+                url: item.url,
+                isBookmarked: item.isBookmarked || false
+              }
+            });
+          }
+        });
+        
+        setSpaceContent(contentBySpace);
+      } catch (error) {
+        console.error('Error loading content:', error);
+      }
+    };
```
262-332 and 46-125 are the same exact code except for the error message

DrewProkarym commented on frontend/src/components/ContentCard.tsx:
```diff
@@ -97,11 +112,21 @@ export function ContentCard({ type, content, onToggleBookmark }: ContentCardProp
         return (
-          <div className="p-3 bg-white dark:bg-white/5 rounded-lg border border-gray-200 dark:border-white/10 hover:bg-gray-50 dark:hover:bg-white/8 hover:border-gray-300 dark:hover:border-white/20 hover:scale-[1.02] transition-all duration-300 shadow-md hover:shadow-lg">
-            <h4 className="text-gray-900 dark:text-white font-semibold text-sm mb-1.5">{content.title}</h4>
-            <p className="text-gray-600 dark:text-white/60 text-xs leading-relaxed mb-2">{content.text}</p>
-            {content.domain && (
-              <p className="text-gray-400 dark:text-white/40 text-xs">{content.domain}</p>
+          <div className="border-l-3 border-white/50 pl-4">
+            <h4 className="text-white font-medium mb-2">{content.title}</h4>
+            {content.url && (
+              <a 
+                href={content.url} 
+                target="_blank" 
+                rel="noopener noreferrer"
+                className="text-blue-400 hover:text-blue-300 text-sm underline block mb-2"
+                onClick={(e) => e.stopPropagation()}
+              >
+                {content.url}
```
@Unicyclop this is valid, please fix it. Anything the user can provide must be sanitized.

DrewProkarym commented on frontend/src/components/MediaUploadDialog.tsx:
```diff
@@ -0,0 +1,314 @@
+      });
+      
+      // Create media using backend API
+      const result = await api.createMedia({
```
@Unicyclop please fix this. It only appears in the two markdown files for documentation so it's a simple fix.

DrewProkarym commented on frontend/src/components/MediaUploadDialog.tsx:
```diff
@@ -0,0 +1,314 @@
+        reader.readAsDataURL(mediaFile);
+      });
+      
+      console.log("Uploading media:", {
+        filename: mediaFile.name,
+        file_size: mediaFile.size,
+        container_id: currentSpaceId
+      });
+      
+      // Create media using backend API
```
I think it is confused by the markdown files made as part of the PR. Ignore this, but please fix the files as I said in the conversation right below this.

Unicyclop committed:
Fix code quality issues from review

- Fixed variable naming consistency in FloatingMenu (filtered* -> updated*)
- Added MIME-extension validation to EditContentDialog for security
- Fixed async test assertion using waitFor instead of setTimeout
- Removed unused imports and variables

DrewProkarym commented on frontend/src/components/FloatingMenu.tsx:
```diff
@@ -196,28 +135,117 @@ export function FloatingMenu({ currentSpaceId, currentUserId, onContentAdded, on
+          ...item,
+          itemType: 'post',
+          displayTitle: item.title || 'Text Post',
+          displayDescription: item.content?.substring(0, 50) + '...' || ''
```
@Unicyclop please commit this suggestion

DrewProkarym commented on frontend/src/App.tsx:
```diff
@@ -31,13 +33,307 @@ export default function App() {
+      try {
+        // Load all content types from localStorage
+        const posts = JSON.parse(localStorage.getItem('ss_posts') || '[]');
+        const media = JSON.parse(localStorage.getItem('ss_media') || '[]');
+        const bookmarks = JSON.parse(localStorage.getItem('ss_bookmarks') || '[]');
```
Outdated

DrewProkarym commented:
The backend unit test ""insertMediaToDatabase rejects too-large base64" at backend\src\services\__tests__\mediaServices.test.js:250 is failing because the "too large file size" is 21MB but this is now allowed. Increase this to 60MB and it should pass.

Unicyclop committed:
Fix review issues and backend test

- Fixed upload path logic for Docker (/app/uploads instead of /app/backend/uploads)
- Updated backend test for 50MB file size limit (was 21MB, now 51MB)
- Added URL validation in ContentCard to prevent XSS (javascript:/data: URLs)
- Fixed substring logic in FloatingMenu to check length before adding ellipsis
- Replaced window.location.reload() with callback for better UX
- Updated .htaccess to Apache 2.4+ syntax (Require instead of Order/Deny/Allow)
- Fixed SECURITY.md documentation for upload directory path
- All 60 backend tests passing

Unicyclop committed:
Fix file size and MIME type inconsistencies across codebase

- Update EditContentDialog.tsx: 10MB50MB, add videos/BMP/SVG/JSON support
- Update SECURITY.md: Document 50MB limit and expanded file types
- Update MERGE_SUMMARY.md: Correct 20MB references to 50MB
- Update mediaServices.test.js: Increase test file size 51MB60MB
- Delete unused secureFileUpload.js (453 lines, no imports found)
- All 60 backend tests passing, 114 frontend tests passing
- File upload limits now consistent at 50MB everywhere

DrewProkarym committed:
Revert docker file

DrewProkarym committed:
fix init.sql

cfurley reviewed (approved):
Very good job, this is the foundation of the media for second space!
